dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -51,40 +52,199 @@ public Pattern getParameterPattern() {
+            }
+        }
+
+        abstract protected ResultSet getColumns() throws SQLException;
+
+        protected DbParameterAccessor createDbParamAccessor(String paramName, ResultSet rs, Direction dir)
+                throws SQLException {
+            String paramTypeName = rs.getString("TYPE_NAME");
+            return createDbParameterAccessor(paramName, dir,
+                           typeMapper.getJDBCSQLTypeForDBType(paramTypeName),
+                           getJavaClass(paramTypeName), rs.getInt("ORDINAL_POSITION") - 1);
+        }
+
+        abstract protected DbParameterAccessor createColumnAccessor(String paramName, ResultSet rs)
+                throws SQLException;
+
+        public Map<String, DbParameterAccessor> getParams() throws SQLException {
+            try (ResultSet rs = getColumns()) {
+                while (rs.next()) {
+                    String paramName = defaultIfNull(rs.getString("COLUMN_NAME"), "");
+                    allParams.put(NameNormaliser.normaliseName(paramName),
+                                  createColumnAccessor(paramName, rs));
+                }
             }
-            rs.close();
             return allParams;
         }
     }
 
-    public Map<String, DbParameterAccessor> getAllProcedureParameters(
-            String procName) throws SQLException {
-        throw new UnsupportedOperationException();
+    private class DatabaseTable extends DatabaseObject {
+
+        private DatabaseTable(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected ResultSet getObjectMetaData() throws SQLException {
+            return getDbMetaData().getTables(null, name.getSchemaName(), name.getObjectName(), null);
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getDbMetaData().getColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        protected DbParameterAccessor createColumnAccessor(String paramName, ResultSet rs)
+                throws SQLException {
+            return createDbParamAccessor(paramName, rs, Direction.INPUT);
+        }
+    }
+
+    abstract private class DatabaseExecutable extends DatabaseObject {
+
+        private DatabaseExecutable(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        abstract protected Direction columnTypeToDir(int columnType);
+
+        protected Direction getParamDirection(ResultSet rs) throws SQLException {
+            Direction paramDir = columnTypeToDir(rs.getInt("COLUMN_TYPE"));
+            if (paramDir == null) {
+                throw new SQLException("Invalid parameter type");
+            }
+            return paramDir;
+        }
+
+        protected DbParameterAccessor createColumnAccessor(String paramName, ResultSet rs)
+                throws SQLException {
+            return createDbParamAccessor(paramName, rs, getParamDirection(rs));
```

Comment:
> Do you mean you think it's OK to have a dummy one? Or it OK to remove it (as it is right now)?

In general it's OK to have unused parameter in some child classes: since we try to abstract things in a more general method.


> The notion of Direction seems contrived and a bit ugly for tables. I can only think DbFit was originally developed with stored procs in mind and tables/views/queries were added later.

That's a a good point. Basically the direction makes sense for SQL queries (input and output parameters). It's possible e.g. for insert statements to have output parameters for auto-generated values (by triggers, auto-increments, etc.).
The thing is that we're using `DbParameterAccessor` both for describing database meta-data (parameters or columns), and also for binding to SQL parameters. It might make sense to split.

Anyway - it's broader topic. Maybe a separate issue can be created to think about such re-design.